<a href="https://colab.research.google.com/github/anandvimal/deeplearning-experiments/blob/master/mnist_keras_1_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Experiment 1.1 MNIST

this experiment is iteration from deep learning with keras book chapter 1

In last model (mnist_keras_1.0) we used 200 epocs and few layers. This time we will try to use less epochs but bigger net (adding more layers dense+activation). so our neural net can learn more with less epochs.

In [0]:
#import all libs needed
from __future__ import print_function
import numpy as np
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers.core import Dense, Activation
from keras.optimizers import SGD
from keras.utils import np_utils

np.random.seed(1671) #this is so our random numbers are same to get same random weights and get similar results to original author. 
#removing above line can have slightly different results in accuracy and loss in the end but overall results will be similar.

In [0]:
NB_EPOCH = 20
BATCH_SIZE = 128
VERBOSE = 1
NB_CLASSES = 10
OPTIMIZER = SGD()
N_HIDDEN = 128
VALIDATION_SPLIT=0.2


In [0]:
#shuffle and split data
(X_train, Y_train), (X_test, Y_test) = mnist.load_data()

In [0]:
RESHAPED = 784 #28*28=784

X_train = X_train.reshape(60000, RESHAPED)
X_test = X_test.reshape(10000, RESHAPED)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')


In [0]:
#normalize
X_train /= 255
X_test /= 255

print(X_train.shape[0], 'train samples')
print(X_test.shape[0], 'test samples')

60000 train samples
10000 test samples


In [0]:
#one hot encoding of labels
Y_train = np_utils.to_categorical(Y_train, NB_CLASSES)
Y_test = np_utils.to_categorical(Y_test, NB_CLASSES)

In [0]:
#model
model = Sequential()
model.add(Dense(N_HIDDEN, input_shape=(RESHAPED,)))
model.add(Activation('relu'))
model.add(Dense(N_HIDDEN))
model.add(Activation('relu'))
model.add(Dense(NB_CLASSES))     #10 hidden layers because we have 10 classes to categorize.
model.add(Activation('softmax'))  
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_7 (Dense)              (None, 128)               100480    
_________________________________________________________________
activation_7 (Activation)    (None, 128)               0         
_________________________________________________________________
dense_8 (Dense)              (None, 128)               16512     
_________________________________________________________________
activation_8 (Activation)    (None, 128)               0         
_________________________________________________________________
dense_9 (Dense)              (None, 10)                1290      
_________________________________________________________________
activation_9 (Activation)    (None, 10)                0         
Total params: 118,282
Trainable params: 118,282
Non-trainable params: 0
_________________________________________________________________


In [0]:
model.compile(loss='categorical_crossentropy',
             optimizer=OPTIMIZER,
              metrics=['accuracy']
             )

In [0]:
history = model.fit(X_train, Y_train,
                   batch_size=BATCH_SIZE, epochs=NB_EPOCH,
                    verbose=VERBOSE, validation_split=VALIDATION_SPLIT
                   )

Instructions for updating:
Use tf.cast instead.
Train on 48000 samples, validate on 12000 samples
Epoch 1/20
48000/48000 [==============================] - 2s 47us/step - loss: 1.4830 - acc: 0.6231 - val_loss: 0.7584 - val_acc: 0.8286
Epoch 2/20
48000/48000 [==============================] - 3s 59us/step - loss: 0.6049 - acc: 0.8463 - val_loss: 0.4550 - val_acc: 0.8851
Epoch 3/20
48000/48000 [==============================] - 3s 60us/step - loss: 0.4398 - acc: 0.8801 - val_loss: 0.3710 - val_acc: 0.9019
Epoch 4/20
48000/48000 [==============================] - 3s 59us/step - loss: 0.3767 - acc: 0.8952 - val_loss: 0.3322 - val_acc: 0.9081
Epoch 5/20
48000/48000 [==============================] - 3s 60us/step - loss: 0.3415 - acc: 0.9025 - val_loss: 0.3055 - val_acc: 0.9147
Epoch 6/20
48000/48000 [==============================] - 3s 60us/step - loss: 0.3175 - acc: 0.9086 - val_loss: 0.2880 - val_acc: 0.9183
Epoch 7/20
48000/48000 [==============================] - 3s 59us/step - loss: 0

In [0]:
score = model.evaluate(X_test, Y_test, verbose=VERBOSE)

10000/10000 [==============================] - 1s 56us/step


In [0]:
print("test score/loss:", score[0])
print("test accuracy:", score[1])

test score/loss: 0.1860255774706602
test accuracy: 0.9461


by comparing to previous attempt (mnist_keras_1.0) to current one: 

*   our accuracy increased from 92.27% to 94.61%
*   and loss decreased from 0.2773858511567116 to 0.1860

by making the neural net denser (more dense+activation layer) our model performed better even with less no of epochs.